### PR TITLE
Adds support for configuring HTTP Feature Policy

### PR DESCRIPTION
**Pull request by jacobbednarz**
A HTTP feature policy is Yet Another HTTP header for instructing the
browser about which features the application intends to make use of and
to lock down access to others. This is a new security mechanism that
ensures that should an application become compromised or a third party
attempts an unexpected action, the browser will override it and maintain
the intended UX.

**WICG specification: https://wicg.github.io/feature-policy/**

The end result is a HTTP header that looks like the following:

```
Feature-Policy: geolocation 'none'; autoplay https://example.com
```

This will prevent the browser from using geolocation and only allow
autoplay on `https://example.com`. Full feature list can be found over
in the [WICG repository](https://github.com/WICG/feature-policy/blob/master/features.md).

As of today Chrome and Safari [have public support](https://www.chromestatus.com/feature/5694225681219584) for this
functionality with Firefox [working on support](https://bugzilla.mozilla.org/show_bug.cgi?id=1390801) and Edge [still pending
acceptance of the suggestion](https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/33507907-support-feature-policy).

### Examples

Using an initializer

```rb
# config/initializers/feature_policy.rb
Rails.application.config.feature_policy do |f|
  f.geolocation :none
  f.camera      :none
  f.payment     "https://secure.example.com"
  f.fullscreen  :self
end
```

In a controller

```rb
class SampleController < ApplicationController
  def index
    feature_policy do |f|
      f.geolocation "https://example.com"
    end
  end
end
```

Some of you might realise that the HTTP feature policy looks pretty
close to that of a Content Security Policy; and you're right. So much so
that I used the Content Security Policy DSL from #31162 as the starting
point for this change.

This change *doesn't* introduce support for defining a feature policy on
an iframe and this has been intentionally done to split the HTTP header
and the HTML element (`iframe`) support. If this is successful, I'll
look to add that on it's own.

Full documentation on HTTP feature policies can be found at
https://wicg.github.io/feature-policy/. Google [have also published](https://developers.google.com/web/updates/2018/06/feature-policy) a
great in-depth write up of this functionality.

### CODE REVIEW

**Comment by jacobbednarz** [comment]
cc @pixeltrix as you implemented the CSP DSL and this is along the same lines.

File: actionpack/test/dispatch/feature_policy_test.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+end
+
+class FeaturePolicyIntegrationTest < ActionDispatch::IntegrationTest
+  class PolicyController < ActionController::Base
+    feature_policy only: :index do |f|
+      f.gyroscope :none
+    end
+
+    feature_policy only: :sample_controller do |f|
+      f.gyroscope nil
```
Comment (jacobbednarz) [inline actionpack/test/dispatch/feature_policy_test.rb]: I'm not sure how I feel about this API. The idea is that you define a global policy you need to be explicit about overriding it with `nil` to remove it. Thoughts?

**Comment by rafaelfranca** [comment]
This seems to be a very unstable API that may change any time. I don't think it is mature enough to be included in a framework just yet.

**Comment by jeremy** [comment]
Cool! Makes me want an explicitly-experimental API status so app developers can track this as it's developed, but the clear expectation that it may change to track upstream.

**Comment by jacobbednarz** [comment]
I can definitely see your concerns here @rafaelfranca and they are warranted as many of these new browser security features can go through multiple iterations before becoming usable by more than a single browser (I'm looking at your content security policy!).

I really like @jeremy's idea of explicitly calling this out as potentially experimental due to the browser implementations not all solidified yet. I think that will give us a nice middle ground of adding security for those that do want to use this feature while still maintaining some sensibility in the framework itself.

**Comment by rafaelfranca** [comment]
I'm ok with that. My worry is that if they rename a configuration it would require a released of the framework to rename that configuration. Having this is a gem would give more flexibility. I'm happy to keep an official Rails gem for this with the promise to integrate to the framework as soon the API is stable enough.

**Comment by jacobbednarz** [comment]
Rather timely occurrence, this morning Mozilla has announced they will be shipping this to nightly - https://groups.google.com/d/msg/mozilla.dev.platform/w3elPpZlqIE/ZA5cuKOMCAAJ

That just leaves Edge without support (for now) from the majority of browsers.

**Comment by rafaelfranca** [comment]
Sounds good to me them. Let's move to document this PR so we can merge before the pre release.

**Comment by jacobbednarz** [comment]
@rafaelfranca Sure thing! Would you like the documentation in this PR or would you link a separate one for that?

**Comment by jacobbednarz** [comment]
@jeremy @rafaelfranca could either of you point me towards the documentation that you're expecting here and we can get this one over the line?

**Comment by jacobbednarz** [comment]
I've gone ahead and updated this to the latest master as well as adding a CHANGELOG entry for the new DSL. Sing out if you want more documentation in other forms.

File: actionpack/lib/action_controller/metal/feature_policy.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module ActionController #:nodoc:
+  module FeaturePolicy
```
Comment (jeremy) [inline actionpack/lib/action_controller/metal/feature_policy.rb]: Good spot for docs here (https://github.com/rails/rails/blob/6e40b131d2c7208e116d267e4b40dcf621f8b442/actionpack/lib/action_controller/metal/content_security_policy.rb#L5 notwithstanding 😅). Using the changelog entry verbatim could work well!

**Review by jeremy — APPROVED** [review]
🙏🏼

File: actionpack/lib/action_controller/metal/feature_policy.rb (hunk 2 above)
Comment (jacobbednarz) [reply actionpack/lib/action_controller/metal/feature_policy.rb]: Thanks for the pointer @jeremy! I wasn't sure if the intention was for code documentation or something like the Guide documentation. I'll get something added here as a start.

Comment (jacobbednarz) [reply actionpack/lib/action_controller/metal/feature_policy.rb]: Addressed this in 88fb316.

**Comment by jacobbednarz** [comment]
Let me know when/if we're ready here and I'll squash this into a single commit before merging it. (Unless whoever merges is happy squashing-when-merging using the GitHub UI and just take the initial commit)

File: railties/lib/rails/generators/rails/app/templates/config/initializers/feature_policy.rb.tt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Define an application-wide HTTP feature policy. For further
+# information see https://developers.google.com/web/updates/2018/06/feature-policy
+#
+# Rails.application.config.feature_policy do |f|
+#   f.camera      :none
+#   f.gyroscope   :none
+#   f.microphone  :none
+#   f.usb         :none
+#   f.fullscreen  :self
+#   f.payment     :self, "https://secure-example.com"
```
Comment (y-yagi) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/feature_policy.rb.tt]: Unlike `example.com`, `secure-example.com` is a domain that can be used normally. It should not be used as an example.

Comment (jacobbednarz) [reply railties/lib/rails/generators/rails/app/templates/config/initializers/feature_policy.rb.tt]: Good catch. While this is only an example for configuration, we should definitely avoid using real world domains and instead restrict to using using the [reserved domains](https://tools.ietf.org/html/rfc2606#section-3). I'll open a PR to address.

**Pull request merged**